terraform/treasury_generation_lambda_functions.tf

```diff
@@ -0,0 +1,174 @@
+
+module "lambda_function-subrecipientTreasuryReportGen" {
+  source  = "terraform-aws-modules/lambda/aws"
+  version = "6.5.0"
+
+  // Metadata
+  function_name = "${var.namespace}-subrecipientTreasuryReportGen"
+  description   = "Generates subrecipients file for treasury report when called by step function."
+
+  // Networking
+  vpc_subnet_ids         = null
+  vpc_security_group_ids = null
+  attach_network_policy  = false
+
+  // Permissions
+  role_permissions_boundary         = local.permissions_boundary_arn
+  attach_cloudwatch_logs_policy     = true
+  cloudwatch_logs_retention_in_days = var.log_retention_in_days
+  attach_policy_jsons               = length(local.lambda_default_execution_policies) > 0
+  number_of_policy_jsons            = length(local.lambda_default_execution_policies)
+  policy_jsons                      = local.lambda_default_execution_policies
+  attach_policy_statements          = true
+  policy_statements = {
+    AllowDownloadSubrecipientsFile = {
+      effect = "Allow"
+      actions = [
+        "s3:GetObject",
+        "s3:HeadObject",
+      ]
+      resources = [
+        # Path: /{organization_id}/{reporting_period_id}/subrecipients
+        "${module.reporting_data_bucket.bucket_arn}/*/*/subrecipients",
+      ]
+    }
+    AllowUploadCSVReport = {
+      effect = "Allow"
+      actions = [
+        "s3:PutObject"
+      ]
+      resources = [
+        # Path: /treasuryreports/{organization_id}/{reporting_period_id}/{userId}/CPFSubrecipientTemplate.csv
+        "${module.reporting_data_bucket.bucket_arn}/treasuryReports/*/*/*/CPFSubrecipientTemplate.csv",
+      ]
+    }
+  }
+
+  // Artifacts
+  publish        = true
+  create_package = false
+  s3_existing_package = {
+    bucket = aws_s3_object.lambda_artifact-python.bucket
+    key    = aws_s3_object.lambda_artifact-python.key
+  }
+
+  // Runtime
+  handler       = var.datadog_enabled ? local.datadog_lambda_py_handler : "src.functions.subrecipient_treasury_report_gen.handle"
+  runtime       = var.lambda_py_runtime
+  architectures = [var.lambda_arch]
+  layers        = local.lambda_py_layer_arns
+  timeout       = 60 # 1 minute, in seconds
+  memory_size   = 512
+  environment_variables = merge(local.lambda_default_environment_variables, {
+    DD_LAMBDA_HANDLER = "src.functions.subrecipient_treasury_report_gen.handle"
+    DD_LOGS_INJECTION = "true"
+  })
+
+  allowed_triggers = {
+    StepFunctionTrigger = {
+      principal  = "states.amazonaws.com"
+      source_arn = module.treasury_generation_step_function.state_machine_arn
+    }
+
+  }
+}
+
+module "lambda_function-treasuryReportGeneration" {
+  for_each = toset(["1A", "1B", "1C"])
```

Comment:
Given the extreme similarity between each of the Lambda functions generated by this module definition, I think it would be better to eliminate the `for_each` and define a single Lambda function for report generation. 

Unless there's a clear operational benefit for having distinct functions (or clear detriment to having them combined), I think merging is justified given that each Lambda in the `for_each` shares the following criteria:
1. Invocation source(s)
2. Execution policy
3. Handler callable (NB: inspecting a payload value feels like a more appropriate way to guide 1A/B/C selection than an env var)

Reply:
Keeping this lambda the same makes sense to me. The advantage I saw was mainly for observability in datadog. By filtering by function ARN it is easy to know whether the Project 1A/1B/1C files are failing to generate as opposed to adding a secondary filter on payload.

Since we can always add additional arguments to the log-context I'm good with making this change and simplifying this setup to have a single lambda.

Reply:
I can see the need for different capacity requirements. I assume there are more of one type of project to generate over time, but I don't have any data at the moment to base this off-of so we can always separate it out when we notice this in the future.

Reply:
> I can see the need for different capacity requirements. I assume there are more of one type of project to generate over time, but I don't have any data at the moment to base this off-of so we can always separate it out when we notice this in the future.

FWIW, I think it will be a while before we're scaled to the point where we'll need to tune concurrency for these Lambda functions, the Lambda + SFN services should be able to work out scheduling executions on their own to a relatively high degree of scale.